File: android/app/src/main/java/com/wayfare/app/ui/TripsScreen.kt
package com.wayfare.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Add
import androidx.compose.material.icons.outlined.Explore
import androidx.compose.material.icons.automirrored.outlined.Logout
import androidx.compose.material.icons.outlined.ConfirmationNumber
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material.icons.outlined.Refresh
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.wayfare.app.AppContainer
import com.wayfare.app.core.TripPhase
import com.wayfare.app.core.tripPhase
import com.wayfare.app.feature.TripsViewModel
import kotlinx.coroutines.launch

@Composable
fun TripsScreen(
    viewModel: TripsViewModel,
    container: AppContainer,
    onOpenTrip: (String) -> Unit,
    onSignOut: () -> Unit,
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val pendingInvite by container.preferences.pendingInvite.collectAsState(initial = null)
    var showTripForm by remember { mutableStateOf(false) }
    var showJoin by remember { mutableStateOf(false) }
    var inviteSeed by remember { mutableStateOf("") }
    var showMenu by remember { mutableStateOf(false) }

    LaunchedEffect(pendingInvite) {
        pendingInvite?.let {
            inviteSeed = it
            showJoin = true
        }
    }

    OnResume(viewModel::onResume)

    Scaffold(containerColor = CanvasWhite) { insets ->
        PullToRefreshBox(
            isRefreshing = state.refreshing,
            onRefresh = viewModel::refresh,
            modifier = Modifier.fillMaxSize().padding(insets),
        ) {
            LazyColumn(
                Modifier.fillMaxSize(),
                // Photographs read as distinct objects only if the canvas between
                // them is wide enough to be read as canvas.
                verticalArrangement = Arrangement.spacedBy(24.dp),
            ) {
                // Top nav: wordmark, circular controls, one hairline underneath.
                item {
                    Column {
                        Row(
                            Modifier.fillMaxWidth().padding(horizontal = 24.dp).padding(top = 8.dp, bottom = 12.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Brand(Modifier.weight(1f))
                            CircleIconButton(Icons.Outlined.ConfirmationNumber, "Join a trip") { showJoin = true }
                            Spacer(Modifier.size(8.dp))
                            Box {
                                CircleIconButton(Icons.Outlined.Person, "Account") { showMenu = true }
                                DropdownMenu(expanded = showMenu, onDismissRequest = { showMenu = false }) {
                                    DropdownMenuItem(
                                        text = { Text("Refresh") },
                                        leadingIcon = { Icon(Icons.Outlined.Refresh, null) },
                                        onClick = { showMenu = false; viewModel.refresh() },
                                    )
                                    DropdownMenuItem(
                                        text = { Text("Sign out") },
                                        leadingIcon = { Icon(Icons.AutoMirrored.Outlined.Logout, null) },
                                        onClick = { showMenu = false; onSignOut() },
                                    )
                                }
                            }
                        }
                        HairlineDivider()
                    }
                }
                item {
                    Column(Modifier.padding(horizontal = 24.dp)) {
                        Text("Your trips", style = MaterialTheme.typography.displaySmall)
                        Text(
                            if (state.trips.isEmpty()) "Your first journey is waiting."
                            else "${state.trips.size} ${if (state.trips.size == 1) "trip" else "trips"}. All your plans, in one place.",
                            Modifier.padding(top = 6.dp, bottom = 20.dp),
                            color = Ash,
                            style = MaterialTheme.typography.bodyLarge,
                        )
                        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                            PrimaryButton("New trip", Modifier.weight(1f), icon = Icons.Outlined.Add) { showTripForm = true }
                            SecondaryButton("Join friend", Modifier.weight(1f)) { showJoin = true }
                        }
                    }
                }
                state.error?.let { message ->
                    item { Notice(message, true, Modifier.padding(horizontal = 24.dp)) }
                }
                if (state.loading) {
                    item {
                        Column(Modifier.fillMaxWidth().padding(48.dp), horizontalAlignment = Alignment.CenterHorizontally) {
                            CircularProgressIndicator(color = Rausch)
                        }
                    }
                } else if (state.trips.isEmpty()) {
                    item {
                        Column(
                            Modifier.fillMaxWidth().padding(horizontal = 24.dp)
                                .clip(RoundedCornerShape(14.dp)).background(SoftCloud).padding(32.dp),
                            horizontalAlignment = Alignment.CenterHorizontally,
                        ) {
                            Box(
                                Modifier.size(64.dp).clip(CircleShape).background(CanvasWhite),
                                contentAlignment = Alignment.Center,
                            ) { Icon(Icons.Outlined.Explore, null, tint = Rausch, modifier = Modifier.size(28.dp)) }
                            Text(
                                "Good trips start here",
                                Modifier.padding(top = 18.dp),
                                style = MaterialTheme.typography.titleLarge,
                            )
                            Text(
                                "Add a trip, set a budget, and log each cost as it lands.",
                                Modifier.padding(top = 6.dp),
                                color = Ash,
                                style = MaterialTheme.typography.bodyMedium,
                                textAlign = TextAlign.Center,
                            )
                            TextButton(onClick = { showJoin = true }, Modifier.padding(top = 8.dp)) {
                                Text("I have an invite code", color = Ink, style = MaterialTheme.typography.labelMedium)
                            }
                        }
                    }
                } else {
                    val sections = listOf(
                        "Active trips" to state.trips.filter { tripPhase(it.trip) is TripPhase.Active },
                        "Upcoming trips" to state.trips.filter { tripPhase(it.trip) is TripPhase.Upcoming },
                        "Dates open" to state.trips.filter { tripPhase(it.trip) is TripPhase.Undated },
                        "Past trips" to state.trips.filter { tripPhase(it.trip) is TripPhase.Past },
                    )
                    sections.forEach { (title, trips) ->
                        if (trips.isNotEmpty()) {
                            item(key = title) {
                                Row(
                                    Modifier.fillMaxWidth().padding(horizontal = 24.dp).padding(top = 8.dp),
                                    verticalAlignment = Alignment.CenterVertically,
                                ) {
                                    Text(title, style = MaterialTheme.typography.headlineSmall)
                                    Text(
                                        trips.size.toString(),
                                        Modifier.padding(start = 8.dp),
                                        color = Ash,
                                        style = MaterialTheme.typography.bodyLarge,
                                    )
                                }
                            }
                            items(trips, key = { it.trip.id }) { summary ->
                                ListingCard(
                                    summary = summary,
                                    coverUrl = container.repository.coverUrl(summary.trip.coverPath),
                                    modifier = Modifier.padding(horizontal = 24.dp).animateItem(),
                                    onClick = { onOpenTrip(summary.trip.id) },
                                )
                            }
                        }
                    }
                }
                item { Spacer(Modifier.height(28.dp)) }
            }
        }
    }

    if (showTripForm) TripFormDialog(
        onDismiss = { showTripForm = false },
        onSave = viewModel::createTrip,
    )
    if (showJoin) JoinTripDialog(
        initialCode = inviteSeed,
        onDismiss = { showJoin = false },
        onJoin = { code ->
            viewModel.joinTrip(code).onSuccess { tripId ->
                container.preferences.setPendingInvite(null)
                showJoin = false
                inviteSeed = ""
                onOpenTrip(tripId)
            }
        },
    )
}

@Composable
private fun JoinTripDialog(
    initialCode: String,
    onDismiss: () -> Unit,
    onJoin: suspend (String) -> Result<*>,
) {
    var code by rememberSaveable(initialCode) { mutableStateOf(initialCode.uppercase().take(8)) }
    var busy by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    AlertDialog(
        onDismissRequest = onDismiss,
        shape = RoundedCornerShape(14.dp),
        title = { Text("Join a trip", style = MaterialTheme.typography.headlineSmall) },
        text = {
            Column {
                Text(
                    "Enter the eight-character code your travel companion sent you.",
                    color = Ash,
                    style = MaterialTheme.typography.bodyMedium,
                )
                OutlinedTextField(
                    code, { code = it.uppercase().filter(Char::isLetterOrDigit).take(8) },
                    Modifier.fillMaxWidth().padding(top = 16.dp), label = { Text("Invite code") }, singleLine = true,
                    shape = RoundedCornerShape(8.dp),
                    colors = wayfareFieldColors(),
                )
                error?.let { Notice(it, true, Modifier.padding(top = 12.dp)) }
            }
        },
        confirmButton = {
            TextButton(enabled = !busy, onClick = {
                if (code.length != 8) error = "Enter all eight characters."
                else scope.launch {
                    busy = true
                    onJoin(code).onFailure { error = it.message ?: "That code did not work." }
                    busy = false
                }
            }) { Text(if (busy) "Joining…" else "Join", color = Rausch, style = MaterialTheme.typography.labelMedium) }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text("Cancel", color = Ink, style = MaterialTheme.typography.labelMedium) } },
    )
}

/**
 * Inputs sit on white behind a hairline border and switch to Ink on focus —
 * the system never tints a field with the accent.
 */
@Composable
fun wayfareFieldColors() = OutlinedTextFieldDefaults.colors(
    focusedBorderColor = Ink,
    unfocusedBorderColor = Hairline,
    errorBorderColor = ErrorRed,
    focusedTextColor = Charcoal,
    unfocusedTextColor = Ink,
    focusedLabelColor = Ash,
    unfocusedLabelColor = Ash,
    errorLabelColor = ErrorRed,
    cursorColor = Ink,
    focusedContainerColor = CanvasWhite,
    unfocusedContainerColor = CanvasWhite,
)

File: ios/Wayfare/TripsScreen.swift
import SwiftUI
import WayfareCore

struct TripsScreen: View {
  @Environment(AppStore.self) private var store
  @State private var path: [String] = []
  @State private var newTrip = false
  @State private var joining = false
  @State private var signout = false
  @State private var recovery = false

  private var sections: [(String, [Trip])] {
    ["Active trips", "Upcoming trips", "Dates open", "Past trips"].map { title in
      (
        title,
        store.trips.filter { trip in
          switch tripPhase(trip) {
          case .active: title == "Active trips"
          case .upcoming: title == "Upcoming trips"
          case .undated: title == "Dates open"
          case .past: title == "Past trips"
          }
        }
      )
    }
  }

  var body: some View {
    NavigationStack(path: $path) {
      ScrollView {
        // Photographs read as distinct objects only if the canvas between them
        // is wide enough to be read as canvas.
        LazyVStack(alignment: .leading, spacing: 24) {
          topNav
          VStack(alignment: .leading, spacing: 6) {
            Text("Your trips").typeStyle(.displaySmall)
            Text(
              store.trips.isEmpty
                ? "Your first journey is waiting."
                : "\(store.trips.count) \(store.trips.count == 1 ? "trip" : "trips"). All your plans, in one place."
            ).typeStyle(.bodyLarge).foregroundStyle(Palette.ash)
          }.padding(.horizontal, 24)
          ViewThatFits(in: .horizontal) {
            HStack(spacing: 12) { tripActions }
            VStack(spacing: 12) { tripActions }
          }.padding(.horizontal, 24)
          if let notice = store.notice { Notice(text: notice).padding(.horizontal, 24) }
          if store.expenses.contains(where: { $0.syncState != .synced }) {
            SecondaryButton(title: "Review unsynced expenses", pill: true) { recovery = true }
              .padding(.horizontal, 24)
          }
          if store.loading && store.trips.isEmpty {
            TripLoadingSkeleton().padding(.horizontal, 24)
          } else if store.trips.isEmpty {
            emptyState.padding(.horizontal, 24)
          }
          ForEach(sections, id: \.0) { title, trips in
            if !trips.isEmpty {
              HStack(spacing: 8) {
                Text(title).typeStyle(.headlineSmall)
                Text("\(trips.count)").typeStyle(.bodyLarge).foregroundStyle(Palette.ash)
              }.padding(.horizontal, 24).padding(.top, 8)
              ForEach(trips) { trip in
                NavigationLink(value: trip.id) { listingCard(trip) }
                  .buttonStyle(.plain).padding(.horizontal, 24)
              }
            }
          }
          Spacer(minLength: 12)
        }.frame(maxWidth: 700).frame(maxWidth: .infinity)
      }
      .canvasScreen().refreshable { await store.refresh() }
      .toolbar(.hidden, for: .navigationBar)
      .navigationDestination(for: String.self) { TripDetailScreen(tripId: $0) }
      .sheet(isPresented: $newTrip) { TripForm(trip: Trip(), isNew: true) { path.append($0) } }
      .sheet(isPresented: $joining) { JoinForm { path.append($0) } }
      .sheet(isPresented: $recovery) { UnsyncedScreen() }
      .onChange(of: store.pendingInvite, initial: true) { _, code in
        if !code.isEmpty { joining = true }
      }
      .confirmationDialog("Sign out?", isPresented: $signout, titleVisibility: .visible) {
        Button("Sign out and clear this device", role: .destructive) {
          Task { try? await store.signOut() }
        }
      } message: {
        Text(
          "Cached trips and any unsynced expenses will be removed from this device. Sync pending entries first to keep them."
        )
      }
    }
  }

  /// Top nav: the Rausch wordmark and circular controls.
  private var topNav: some View {
    HStack(spacing: 8) {
      HStack(spacing: 8) {
        Image(systemName: "airplane.departure").font(.system(size: 20, weight: .medium))
        Text("wayfare").typeStyle(.headlineSmall)
      }.foregroundStyle(Palette.rausch)
      Spacer(minLength: 0)
      Menu {
        Button("Refresh", systemImage: "arrow.clockwise") { Task { await store.refresh() } }
        Button("Sign out", systemImage: "rectangle.portrait.and.arrow.right", role: .destructive)
        { signout = true }
      } label: {
        Image(systemName: "person.crop.circle").font(.system(size: 16, weight: .medium))
          .foregroundStyle(Palette.ink).frame(width: 44, height: 44)
          .background(Palette.softCloud, in: Circle())
      }.accessibilityLabel("Account")
    }.padding(.horizontal, 24).padding(.top, 8)
  }

  @ViewBuilder private var tripActions: some View {
    PrimaryButton(title: "New trip", icon: "plus") { newTrip = true }
    SecondaryButton(title: "Join friend") { joining = true }
  }

  private var emptyState: some View {
    VStack(spacing: 0) {
      Image(systemName: "safari").font(.system(size: 28, weight: .light))
        .foregroundStyle(Palette.rausch).frame(width: 64, height: 64)
        .background(Palette.canvas, in: Circle())
      Text("Good trips start here").typeStyle(.titleLarge).padding(.top, 18)
      Text("Add a trip, set a budget, and log each cost as it lands.")
        .typeStyle(.bodyMedium).foregroundStyle(Palette.ash).multilineTextAlignment(.center)
        .padding(.top, 6)
      Button("I have an invite code") { joining = true }
        .typeStyle(.labelMedium).foregroundStyle(Palette.ink).padding(.top, 12)
    }
    .frame(maxWidth: .infinity).padding(32)
    .background(Palette.softCloud, in: RoundedRectangle(cornerRadius: Radius.card))
  }

  /// A photo-led destination card with a native, content-sized overlay.
  private func listingCard(_ trip: Trip) -> some View {
    let expenses = store.expenses.filter { $0.tripId == trip.id }
    let summary = budgetSummary(trip, expenses: expenses)
    let destination = trip.destination?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    let isPrint = trip.coverPath?.hasSuffix("-print.jpg") == true
    // The cover stays untouched apart from the countdown; the numbers ride in a
    // drawer tucked under it. Only covers without lettering need the place
    // named in words.
    return VStack(spacing: 0) {
      TripArtwork(trip: trip, url: store.coverURL(trip.coverPath), aspect: 3 / 2, showStatus: false)
        .accessibilityHidden(true)
        .overlay(alignment: .topLeading) { PhaseBadge(trip: trip).padding(12) }
        .clipShape(RoundedRectangle(cornerRadius: 26, style: .continuous))
        .padding(5)
        .background(Palette.canvas, in: RoundedRectangle(cornerRadius: 31, style: .continuous))
        .shadow(color: .black.opacity(0.08), radius: 12, x: 0, y: 6)
        .zIndex(1)
      HStack(spacing: 12) {
        Group {
          if trip.coverStatus == "pending" {
            Text("Creating your cover…").foregroundColor(Palette.ash)
          } else {
            Text(isPrint ? "" : "\(destination.isEmpty ? trip.name : destination) · ").foregroundColor(Palette.ash)
              + Text(money(summary.spent, trip.currency)).fontWeight(.bold).foregroundColor(Palette.rauschDeep)
              + Text(" spent").foregroundColor(Palette.ash)
          }
        }
        .font(.subheadline).monospacedDigit().lineLimit(1)
        Spacer(minLength: 0)
        Image(systemName: "arrow.up.right").font(.subheadline.weight(.semibold))
          .foregroundStyle(Palette.ink)
      }
      .padding(.horizontal, 18).padding(.top, 26).padding(.bottom, 14)
      .background(Palette.softCloud, in: UnevenRoundedRectangle(bottomLeadingRadius: 20, bottomTrailingRadius: 20, style: .continuous))
      .overlay(UnevenRoundedRectangle(bottomLeadingRadius: 20, bottomTrailingRadius: 20, style: .continuous)
        .stroke(Palette.hairline, lineWidth: 1))
      .padding(.top, -12).padding(.horizontal, 14)
    }
    .foregroundStyle(Palette.ink)
    .accessibilityElement(children: .ignore)
    .accessibilityLabel("\(destination.isEmpty ? trip.name : destination), \(phaseLabel(trip)), \(money(summary.spent, trip.currency)) spent")
    .accessibilityHint("Opens trip details")
  }

}

struct JoinForm: View {
  @Environment(AppStore.self) private var store
  @Environment(\.dismiss) private var dismiss
  @State private var code = ""
  @State private var busy = false
  @State private var error: String?
  let onJoin: (String) -> Void
  var body: some View {
    NavigationStack {
      ScrollView {
        VStack(alignment: .leading, spacing: 16) {
          Text("Enter the eight-character code your travel companion sent you.")
            .typeStyle(.bodyMedium).foregroundStyle(Palette.ash)
          WayfareTextField(label: "Invite code", text: $code, error: error)
            .textInputAutocapitalization(.characters).autocorrectionDisabled()
        }.padding(24).frame(maxWidth: 700).frame(maxWidth: .infinity)
      }.canvasScreen()
        .safeAreaInset(edge: .bottom, spacing: 0) {
          ActionBar { PrimaryButton(title: "Join trip", busy: busy, action: join) }
        }
        .navigationTitle("Join a trip").navigationBarTitleDisplayMode(.inline)
        .toolbar {
          ToolbarItem(placement: .cancellationAction) {
            Button("Cancel") {
              store.pendingInvite = ""
              dismiss()
            }.disabled(busy)
          }
        }
    }.onAppear { code = store.pendingInvite }.interactiveDismissDisabled(busy)
  }

  private func join() {
    let clean = code.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    guard clean.count == 8, clean.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber) })
    else {
      error = "Enter all eight letters and numbers."
      return
    }
    busy = true
    Task {
      defer { busy = false }
      do {
        let id = try await store.joinTrip(code: clean)
        store.pendingInvite = ""
        dismiss()
        onJoin(id)
      } catch { self.error = error.localizedDescription }
    }
  }
}

// Kept outside trip navigation: removed membership must not hide a failed entry.
struct UnsyncedScreen: View {
  @Environment(AppStore.self) private var store
  @Environment(\.dismiss) private var dismiss
  @State private var deleting: Expense?
  @State private var busy = false
  @State private var error: String?
  var body: some View {
    NavigationStack {
      List {
        if let error { Notice(text: error, isError: true) }
        ForEach(store.expenses.filter { $0.syncState != .synced }) { expense in
          VStack(alignment: .leading, spacing: 10) {
            Text(expense.title).typeStyle(.titleMedium)
            Text("Ledger amount: \(money(expense.amount)) · \(expense.spentOn)")
              .typeStyle(.bodyMedium).foregroundStyle(Palette.ash)
            Text(
              expense.syncState == .failed
                ? "Not saved. Excluded from totals." : "Saved on this device. Waiting to sync."
            ).typeStyle(.bodyMedium)
            if let reason = expense.syncError {
              Text(reason).typeStyle(.bodySmall).foregroundStyle(Palette.ash)
            }
            Button("Try again") { run { try await store.retryExpense(expense) } }
              .typeStyle(.labelMedium).foregroundStyle(Palette.ink)
            if expense.syncState == .failed {
              Button("Discard", role: .destructive) { deleting = expense }
                .typeStyle(.labelMedium).foregroundStyle(Palette.errorRed)
            }
          }.padding(.vertical, 8)
        }
      }.canvasScreen().disabled(busy).navigationTitle("Unsynced expenses")
        .toolbar {
          ToolbarItem(placement: .confirmationAction) {
            Button("Done") { dismiss() }.disabled(busy)
          }
        }
        .confirmationDialog(
          "Discard this unsaved expense?",
          isPresented: Binding(get: { deleting != nil }, set: { if !$0 { deleting = nil } }),
          titleVisibility: .visible
        ) {
          if let expense = deleting {
            Button("Discard expense", role: .destructive) {
              run { try await store.discardExpense(expense) }
            }
          }
        }
    }.interactiveDismissDisabled(busy)
  }
  private func run(_ action: @escaping () async throws -> Void) {
    busy = true
    Task {
      defer { busy = false }
      do { try await action() } catch { self.error = error.localizedDescription }
    }
  }
}
